### ISSUE TITLE
A layout using another layout as a base

### ISSUE THREAD

**Issue by elpiel** [cobalt-org/cobalt.rs#942]
Since I'm looking to replace `jekyll` with something that is built with Rust, I realised that it's not currently possible to have a layout that uses another layout.

I tried extracting the frontmatter for the layer and making a look when rendering the document, however, I was unable to make it work.

Is this something you'd like to have in cobalt?
I can open a draft PR if it is so.

**Comment by epage** [comment]
Have you looked into using `include` to use snippets between these layouts?

My preference is to instead keep it simple, relying on composing using `include` but I'm up for hearing use cases to explain why layouts of layouts would be needed.

**Comment by 0xc0000007b** [comment]
> Have you looked into using include to use snippets between these layouts?

I'm not OP but I can attest one of the greatest features of Hugo are layout inheritance. Without it it's not possible to build a DRY website or practice functional programming via composition in the UI.

**Comment by elpiel** [comment]
> Have you looked into using `include` to use snippets between these layouts?
> 
> My preference is to instead keep it simple, relying on composing using `include` but I'm up for hearing use cases to explain why layouts of layouts would be needed.

Yeah, this is what the default layout is doing, but the crates one extends it to add more stuff:

https://github.com/AeroRust/are-we-in-space-yet/blob/933b45db16c2585f68d2ce0c9a03640ad6604366/_layouts/crates.html

**Comment by epage** [comment]
Can you help me understand why nested layouts are such a great feature?

To me
- They make it hard to see whats happening, having to follow up through several layers of layouts
- They seem restrictive in their reuse (all or nothing nesting)
- They make creation of god layouts that handle all cases the easy path because of the above restriction.  Its easier to say "just one more thing" than to refactor.

This is compared to a layout built of snippets where you pick and choose the descriptive reuse elements you want.

The one value I see in layouts is maintaining the balanced start/end together rather than having separate includes for start/end, much like RAII in C++ and Rust or `with` in Python.  I can see us adding support for something that is a mix of Tera [filter blocks](https://github.com/cobalt-org/liquid-rust/issues/321) and [macros](https://github.com/cobalt-org/liquid-rust/issues/322)
```liquid
{% define body content %}
<body>
{{ content }}
</body>
{% enddefine%}

{% body %}
Hello world!
{%endbody %}
```

**Comment by RivenSkaye** [comment]
I have a use case example for nested layouts, which also stems from migrating from Jekyll and using a theme.
- Stuff that's the same everywhere, like the generated CSS file
- Same page layout, doctype, head, etc.
- including things like goatcounter everywhere for reliable stats

I've already broken up the `<head>...</head>` into a separate include, but most of the base template on my blog is applicable regardless of the page. [Repository for reference](https://codeberg.org/RivenSkaye/ramblings/src/branch/master/_layouts/default.html)

It follows the same pattern of deduplication applicable to code, where all of the common stuff is shoved into an outer default template, so changing it in one place keeps it consistent everywhere
